File: src/geometry/geometryValueKernels.ts
import type { ArcDirection } from "../types/geometry";
import type {
  ComputedGeometryValueBezierCurve,
  ComputedGeometryValueOffsetLine,
  ComputedGeometryValueOffsetLineSegment,
  ComputedGeometryValuePolyline,
  ComputedOffsetLine
} from "./evaluationTypes";
import {
  approximateCubicLength,
  cross,
  cubicDerivativeAt,
  cubicPointAt,
  dot,
  EPSILON,
  selectBestBezierFeatureCandidate,
  solveRealQuadratic,
  type BezierLikeSegment
} from "./bezierMath";
import { CIRCLE_EPSILON, degreesToRadians, directedSweepDegrees } from "./evaluateGeometryPrimitives";
import { arcTangentAngles, lineTangentAngles } from "./lineMeasurements";

export type StructuralPoint = { x: number; y: number };

/** Identity-free structural form of the drawable Bezier extreme-point calculation. */
export const bezierExtremePointGeometryKernel = (
  segment: BezierLikeSegment,
  direction: StructuralPoint
): StructuralPoint => {
  const derivativeProjection = (t: number) => dot(cubicDerivativeAt(segment, t), direction);
  const f0 = derivativeProjection(0);
  const fHalf = derivativeProjection(0.5);
  const f1 = derivativeProjection(1);
  const c = f0;
  const a = 2 * (f1 + f0 - 2 * fHalf);
  const b = f1 - f0 - a;
  const candidates = [0, 1].map((t) => ({
    t,
    score: dot(cubicPointAt(segment, t), direction)
  }));

  for (const root of solveRealQuadratic(a, b, c)) {
    if (root > 0 && root < 1) {
      candidates.push({
        t: root,
        score: dot(cubicPointAt(segment, root), direction)
      });
    }
  }

  if (Math.abs(f0) <= EPSILON && Math.abs(fHalf) <= EPSILON && Math.abs(f1) <= EPSILON) {
    candidates.push({
      t: 0.5,
      score: dot(cubicPointAt(segment, 0.5), direction)
    });
  }

  const selected = selectBestBezierFeatureCandidate(candidates);
  return cubicPointAt(segment, selected?.t ?? 0.5);
};

/** Identity-free structural form of the drawable Bezier bulge-point calculation. */
export const bezierBulgePointGeometryKernel = (
  segment: BezierLikeSegment
): StructuralPoint | null => {
  const chord = {
    x: segment.end.x - segment.start.x,
    y: segment.end.y - segment.start.y
  };
  const chordLength = Math.hypot(chord.x, chord.y);
  if (chordLength <= EPSILON) return null;

  const derivativeCross = (t: number) => cross(chord, cubicDerivativeAt(segment, t));
  const q0 = derivativeCross(0);
  const qHalf = derivativeCross(0.5);
  const q1 = derivativeCross(1);
  const c = q0;
  const a = 2 * (q1 + q0 - 2 * qHalf);
  const b = q1 - q0 - a;
  const scoreAt = (t: number) =>
    Math.abs(cross(chord, {
      x: cubicPointAt(segment, t).x - segment.start.x,
      y: cubicPointAt(segment, t).y - segment.start.y
    })) / chordLength;
  const candidates: { t: number; score: number }[] = [];

  for (const root of solveRealQuadratic(a, b, c)) {
    if (root > 0 && root < 1) candidates.push({ t: root, score: scoreAt(root) });
  }
  if (Math.abs(q0) <= EPSILON && Math.abs(qHalf) <= EPSILON && Math.abs(q1) <= EPSILON) {
    candidates.push({ t: 0.5, score: scoreAt(0.5) });
  }

  const selected = selectBestBezierFeatureCandidate(candidates);
  return cubicPointAt(segment, selected?.t ?? 0.5);
};

export type StructuralLine = {
  kind: "line";
  start: StructuralPoint;
  end: StructuralPoint;
  length: number;
  startAngleDeg: number | null;
  endAngleDeg: number | null;
  startTangentAngleDeg: number | null;
  endTangentAngleDeg: number | null;
};

export type StructuralArcLine = {
  kind: "arcLine";
  center: StructuralPoint;
  start: StructuralPoint;
  end: StructuralPoint;
  radius: number;
  startAngleDeg: number;
  endAngleDeg: number;
  startTangentAngleDeg: number;
  endTangentAngleDeg: number;
  sweepAngleDeg: number;
  length: number;
};

export const coordinateGeometryKernel = (x: number, y: number): StructuralPoint => ({ x, y });

export const offsetPointGeometryKernel = (
  from: StructuralPoint,
  dx: number,
  dy: number
): StructuralPoint => ({ x: from.x + dx, y: from.y + dy });

export const polarPointGeometryKernel = (
  from: StructuralPoint,
  angleDeg: number,
  distance: number
): StructuralPoint => {
  const angleRad = degreesToRadians(angleDeg);
  return {
    x: from.x + Math.cos(angleRad) * distance,
    y: from.y + Math.sin(angleRad) * distance
  };
};

export const divisionPointGeometryKernel = (
  start: StructuralPoint,
  end: StructuralPoint,
  placement: { kind: "distance" | "ratio"; value: number }
): StructuralPoint | null => {
  const vector = { x: end.x - start.x, y: end.y - start.y };
  if (placement.kind === "ratio") {
    return {
      x: start.x + vector.x * placement.value,
      y: start.y + vector.y * placement.value
    };
  }
  const length = Math.hypot(vector.x, vector.y);
  if (length <= CIRCLE_EPSILON) return null;
  return {
    x: start.x + (vector.x / length) * placement.value,
    y: start.y + (vector.y / length) * placement.value
  };
};

const identityFreePoint = ({ x, y }: { x: number; y: number }) => ({ x, y });

const identityFreeOffsetSegment = (
  segment: ComputedOffsetLine["segments"][number]
): ComputedGeometryValueOffsetLineSegment => {
  if (segment.kind === "line") {
    return { kind: "line", start: identityFreePoint(segment.start), end: identityFreePoint(segment.end), length: segment.length };
  }
  if (segment.kind === "bezier") {
    return {
      kind: "bezier",
      start: identityFreePoint(segment.start),
      control1: segment.control1,
      control2: segment.control2,
      end: identityFreePoint(segment.end),
      length: segment.length
    };
  }
  return {
    kind: "arc",
    center: identityFreePoint(segment.center),
    start: identityFreePoint(segment.start),
    end: identityFreePoint(segment.end),
    radius: segment.radius,
    startAngleDeg: segment.startAngleDeg,
    sweepAngleDeg: segment.sweepAngleDeg,
    length: segment.length
  };
};

export const offsetLineGeometryValueKernel = (
  line: ComputedOffsetLine
): ComputedGeometryValueOffsetLine => ({
  kind: "offsetLine",
  start: line.start ? identityFreePoint(line.start) : null,
  end: line.end ? identityFreePoint(line.end) : null,
  segments: line.segments.map(identityFreeOffsetSegment),
  closed: line.closed,
  length: line.length,
  startTangentAngleDeg: line.startTangentAngleDeg,
  endTangentAngleDeg: line.endTangentAngleDeg
});

export const segmentGeometryKernel = (start: StructuralPoint, end: StructuralPoint): StructuralLine => ({
  kind: "line",
  start,
  end,
  length: Math.hypot(end.x - start.x, end.y - start.y),
  ...lineTangentAngles(start, end)
});

export const polarLineGeometryKernel = (
  start: StructuralPoint,
  angleDeg: number,
  length: number
): StructuralLine => segmentGeometryKernel(start, polarPointGeometryKernel(start, angleDeg, length));

export const polylineGeometryKernel = (
  points: readonly StructuralPoint[],
  closed: boolean
): ComputedGeometryValuePolyline | null => {
  const minimumPointCount = closed ? 3 : 2;
  if (points.length < minimumPointCount || points.some((point) => !Number.isFinite(point.x) || !Number.isFinite(point.y))) return null;
  const segments = points.slice(0, -1).map((start, index) => {
    const end = points[index + 1]!;
    return { start, end, length: Math.hypot(end.x - start.x, end.y - start.y) };
  });
  const first = points[0]!;
  const last = points.at(-1)!;
  if (closed && Math.hypot(last.x - first.x, last.y - first.y) > CIRCLE_EPSILON) {
    segments.push({ start: last, end: first, length: Math.hypot(first.x - last.x, first.y - last.y) });
  }
  const nonZero = segments.filter((segment) => segment.length > CIRCLE_EPSILON);
  const startTangentAngleDeg = nonZero[0] ? lineTangentAngles(nonZero[0].start, nonZero[0].end).startTangentAngleDeg : null;
  const endTangentAngleDeg = nonZero.at(-1) ? lineTangentAngles(nonZero.at(-1)!.start, nonZero.at(-1)!.end).endTangentAngleDeg : null;
  return {
    kind: "polyline",
    segments,
    closed,
    start: first,
    end: closed ? first : last,
    length: segments.reduce((sum, segment) => sum + segment.length, 0),
    startTangentAngleDeg,
    endTangentAngleDeg
  };
};

export const arcGeometryKernel = (
  center: StructuralPoint,
  radius: number,
  startAngleDeg: number,
  endAngleDeg: number,
  direction: ArcDirection
): StructuralArcLine => {
  const sweepAngleDeg = directedSweepDegrees(startAngleDeg, endAngleDeg, direction);
  const startAngleRad = degreesToRadians(startAngleDeg);
  const endAngleRad = degreesToRadians(endAngleDeg);
  return {
    kind: "arcLine",
    center,
    start: {
      x: center.x + Math.cos(startAngleRad) * radius,
      y: center.y + Math.sin(startAngleRad) * radius
    },
    end: {
      x: center.x + Math.cos(endAngleRad) * radius,
      y: center.y + Math.sin(endAngleRad) * radius
    },
    radius,
    startAngleDeg,
    endAngleDeg,
    ...arcTangentAngles({ startAngleDeg, endAngleDeg, sweepAngleDeg }),
    sweepAngleDeg,
    length: radius * Math.abs(degreesToRadians(sweepAngleDeg))
  };
};

export const throughArcGeometryKernel = (
  point1: StructuralPoint,
  point2: StructuralPoint,
  point3: StructuralPoint,
  startAngleDeg: number,
  endAngleDeg: number
): StructuralArcLine | null => {
  const denominator =
    2 *
    (point1.x * (point2.y - point3.y) +
      point2.x * (point3.y - point1.y) +
      point3.x * (point1.y - point2.y));

  if (Math.abs(denominator) < CIRCLE_EPSILON) return null;

  const point1Squared = point1.x * point1.x + point1.y * point1.y;
  const point2Squared = point2.x * point2.x + point2.y * point2.y;
  const point3Squared = point3.x * point3.x + point3.y * point3.y;
  const centerX =
    (point1Squared * (point2.y - point3.y) +
      point2Squared * (point3.y - point1.y) +
      point3Squared * (point1.y - point2.y)) /
    denominator;
  const centerY =
    (point1Squared * (point3.x - point2.x) +
      point2Squared * (point1.x - point3.x) +
      point3Squared * (point2.x - point1.x)) /
    denominator;
  const radius = Math.hypot(point1.x - centerX, point1.y - centerY);

  if (!Number.isFinite(centerX) || !Number.isFinite(centerY) || !Number.isFinite(radius) || radius <= CIRCLE_EPSILON) {
    return null;
  }
  return arcGeometryKernel({ x: centerX, y: centerY }, radius, startAngleDeg, endAngleDeg, "counterclockwise");
};

export type BezierGeometryIntermediate = {
  point: StructuralPoint;
  angleDeg: number;
  incomingLength: number;
  outgoingLength: number;
};

const bezierHandlePoint = (point: StructuralPoint, angleDeg: number, length: number): StructuralPoint => {
  const angleRad = degreesToRadians(angleDeg);
  return {
    x: point.x + Math.cos(angleRad) * length,
    y: point.y + Math.sin(angleRad) * length
  };
};

export const bezierGeometryKernel = (
  start: StructuralPoint,
  end: StructuralPoint,
  startAngleDeg: number,
  startLength: number,
  endAngleDeg: number,
  endLength: number,
  intermediates: readonly BezierGeometryIntermediate[]
): ComputedGeometryValueBezierCurve | null => {
  const anchors = [start, ...intermediates.map((intermediate) => intermediate.point), end];
  const outgoingHandles = [
    bezierHandlePoint(start, startAngleDeg, startLength),
    ...intermediates.map((intermediate) => bezierHandlePoint(intermediate.point, intermediate.angleDeg, intermediate.outgoingLength))
  ];
  const incomingHandles = [
    ...intermediates.map((intermediate) => bezierHandlePoint(intermediate.point, intermediate.angleDeg + 180, intermediate.incomingLength)),
    bezierHandlePoint(end, endAngleDeg + 180, endLength)
  ];
  const segments: BezierLikeSegment[] = anchors.slice(0, -1).map((anchor, index) => ({
    start: anchor,
    control1: outgoingHandles[index]!,
    control2: incomingHandles[index]!,
    end: anchors[index + 1]!
  }));
  if (segments.some((segment) => Object.values(segment).some((point) => !Number.isFinite(point.x) || !Number.isFinite(point.y)))) return null;
  return {
    kind: "bezierCurve",
    segments,
    length: segments.reduce((sum, segment) => sum + approximateCubicLength(segment), 0)
  };
};

File: rust-evaluator/src/evaluation/geometry_value_kernels.rs
use super::math::{arc_tangent_angles, positive_sweep_degrees, CIRCLE_EPSILON};

#[derive(Debug, Clone, Copy, PartialEq)]
pub(crate) struct StructuralPoint {
    pub(crate) x: f64,
    pub(crate) y: f64,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub(crate) struct StructuralSegment {
    pub(crate) start: StructuralPoint,
    pub(crate) end: StructuralPoint,
    pub(crate) length: f64,
    pub(crate) start_angle_deg: Option<f64>,
    pub(crate) end_angle_deg: Option<f64>,
    pub(crate) start_tangent_angle_deg: Option<f64>,
    pub(crate) end_tangent_angle_deg: Option<f64>,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub(crate) struct StructuralArcLine {
    pub(crate) center: StructuralPoint,
    pub(crate) start: StructuralPoint,
    pub(crate) end: StructuralPoint,
    pub(crate) radius: f64,
    pub(crate) start_angle_deg: f64,
    pub(crate) end_angle_deg: f64,
    pub(crate) start_tangent_angle_deg: f64,
    pub(crate) end_tangent_angle_deg: f64,
    pub(crate) sweep_angle_deg: f64,
    pub(crate) length: f64,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub(crate) struct StructuralPolylineSegment {
    pub(crate) start: StructuralPoint,
    pub(crate) end: StructuralPoint,
    pub(crate) length: f64,
}

#[derive(Debug, Clone, PartialEq)]
pub(crate) struct StructuralPolyline {
    pub(crate) segments: Vec<StructuralPolylineSegment>,
    pub(crate) closed: bool,
    pub(crate) start: StructuralPoint,
    pub(crate) end: StructuralPoint,
    pub(crate) length: f64,
    pub(crate) start_tangent_angle_deg: Option<f64>,
    pub(crate) end_tangent_angle_deg: Option<f64>,
}

pub(crate) fn coordinate_geometry_kernel(x: f64, y: f64) -> StructuralPoint {
    StructuralPoint { x, y }
}

pub(crate) fn offset_point_geometry_kernel(
    from: StructuralPoint,
    dx: f64,
    dy: f64,
) -> StructuralPoint {
    StructuralPoint {
        x: from.x + dx,
        y: from.y + dy,
    }
}

fn angle_from_to(start: StructuralPoint, end: StructuralPoint) -> Option<f64> {
    let dx = end.x - start.x;
    let dy = end.y - start.y;
    let length = dx.hypot(dy);
    (length > CIRCLE_EPSILON).then(|| {
        let normalized = dy.atan2(dx).to_degrees().rem_euclid(360.0);
        if normalized.abs() < CIRCLE_EPSILON || (360.0 - normalized).abs() < CIRCLE_EPSILON {
            0.0
        } else {
            normalized
        }
    })
}

pub(crate) fn segment_geometry_kernel(
    start: StructuralPoint,
    end: StructuralPoint,
) -> StructuralSegment {
    let start_angle_deg = angle_from_to(start, end);
    let end_angle_deg = angle_from_to(end, start);
    StructuralSegment {
        start,
        end,
        length: (end.x - start.x).hypot(end.y - start.y),
        start_angle_deg,
        end_angle_deg,
        start_tangent_angle_deg: start_angle_deg,
        end_tangent_angle_deg: end_angle_deg,
    }
}

pub(crate) fn polyline_geometry_kernel(
    points: &[StructuralPoint],
    closed: bool,
) -> Option<StructuralPolyline> {
    let minimum = if closed { 3 } else { 2 };
    if points.len() < minimum
        || points
            .iter()
            .any(|point| !point.x.is_finite() || !point.y.is_finite())
    {
        return None;
    }
    let mut segments = points
        .windows(2)
        .map(|pair| StructuralPolylineSegment {
            start: pair[0],
            end: pair[1],
            length: (pair[1].x - pair[0].x).hypot(pair[1].y - pair[0].y),
        })
        .collect::<Vec<_>>();
    let first = *points.first()?;
    let last = *points.last()?;
    if closed && (last.x - first.x).hypot(last.y - first.y) > CIRCLE_EPSILON {
        segments.push(StructuralPolylineSegment {
            start: last,
            end: first,
            length: (first.x - last.x).hypot(first.y - last.y),
        });
    }
    let first_nonzero = segments
        .iter()
        .find(|segment| segment.length > CIRCLE_EPSILON);
    let last_nonzero = segments
        .iter()
        .rev()
        .find(|segment| segment.length > CIRCLE_EPSILON);
    Some(StructuralPolyline {
        length: segments.iter().map(|segment| segment.length).sum(),
        start_tangent_angle_deg: first_nonzero
            .and_then(|segment| angle_from_to(segment.start, segment.end)),
        end_tangent_angle_deg: last_nonzero
            .and_then(|segment| angle_from_to(segment.end, segment.start)),
        segments,
        closed,
        start: first,
        end: if closed { first } else { last },
    })
}

pub(crate) fn direct_arc_geometry_kernel(
    center: StructuralPoint,
    radius: f64,
    start_angle_deg: f64,
    end_angle_deg: f64,
    direction: &str,
) -> StructuralArcLine {
    let sweep_angle_deg = if direction == "clockwise" {
        -positive_sweep_degrees(end_angle_deg, start_angle_deg)
    } else {
        positive_sweep_degrees(start_angle_deg, end_angle_deg)
    };
    let (start_tangent_angle_deg, end_tangent_angle_deg) =
        arc_tangent_angles(start_angle_deg, end_angle_deg, sweep_angle_deg);
    let start_angle_rad = start_angle_deg.to_radians();
    let end_angle_rad = end_angle_deg.to_radians();
    StructuralArcLine {
        center,
        start: StructuralPoint {
            x: center.x + start_angle_rad.cos() * radius,
            y: center.y + start_angle_rad.sin() * radius,
        },
        end: StructuralPoint {
            x: center.x + end_angle_rad.cos() * radius,
            y: center.y + end_angle_rad.sin() * radius,
        },
        radius,
        start_angle_deg,
        end_angle_deg,
        start_tangent_angle_deg,
        end_tangent_angle_deg,
        sweep_angle_deg,
        length: radius * sweep_angle_deg.to_radians().abs(),
    }
}

pub(crate) fn through_arc_geometry_kernel(
    point1: StructuralPoint,
    point2: StructuralPoint,
    point3: StructuralPoint,
    start_angle_deg: f64,
    end_angle_deg: f64,
) -> Option<StructuralArcLine> {
    let denominator = 2.0
        * (point1.x * (point2.y - point3.y)
            + point2.x * (point3.y - point1.y)
            + point3.x * (point1.y - point2.y));
    if denominator.abs() < CIRCLE_EPSILON {
        return None;
    }

    let point1_squared = point1.x * point1.x + point1.y * point1.y;
    let point2_squared = point2.x * point2.x + point2.y * point2.y;
    let point3_squared = point3.x * point3.x + point3.y * point3.y;
    let center_x = (point1_squared * (point2.y - point3.y)
        + point2_squared * (point3.y - point1.y)
        + point3_squared * (point1.y - point2.y))
        / denominator;
    let center_y = (point1_squared * (point3.x - point2.x)
        + point2_squared * (point1.x - point3.x)
        + point3_squared * (point2.x - point1.x))
        / denominator;
    let radius = (point1.x - center_x).hypot(point1.y - center_y);

    if !center_x.is_finite()
        || !center_y.is_finite()
        || !radius.is_finite()
        || radius <= CIRCLE_EPSILON
    {
        return None;
    }
    Some(direct_arc_geometry_kernel(
        StructuralPoint {
            x: center_x,
            y: center_y,
        },
        radius,
        start_angle_deg,
        end_angle_deg,
        "counterclockwise",
    ))
}

#[cfg(test)]
mod tests {
    use super::{
        coordinate_geometry_kernel, direct_arc_geometry_kernel, polyline_geometry_kernel,
        segment_geometry_kernel, through_arc_geometry_kernel, StructuralPoint,
    };

    #[test]
    fn structural_segment_is_identity_free_and_deterministic() {
        let start = coordinate_geometry_kernel(0.0, 0.0);
        let end = coordinate_geometry_kernel(3.0, 4.0);
        assert_eq!(
            segment_geometry_kernel(start, end),
            super::StructuralSegment {
                start: StructuralPoint { x: 0.0, y: 0.0 },
                end: StructuralPoint { x: 3.0, y: 4.0 },
                length: 5.0,
                start_angle_deg: Some(53.13010235415598),
                end_angle_deg: Some(233.13010235415598),
                start_tangent_angle_deg: Some(53.13010235415598),
                end_tangent_angle_deg: Some(233.13010235415598),
            }
        );
    }

    #[test]
    fn structural_direct_arc_is_identity_free_and_directional() {
        let arc = direct_arc_geometry_kernel(
            coordinate_geometry_kernel(0.0, 0.0),
            10.0,
            0.0,
            90.0,
            "clockwise",
        );
        assert_eq!(arc.start, StructuralPoint { x: 10.0, y: 0.0 });
        assert!(arc.end.x.abs() < 1e-12);
        assert!((arc.end.y - 10.0).abs() < 1e-12);
        assert_eq!(arc.sweep_angle_deg, -270.0);
        assert_eq!(arc.length, 10.0 * 270.0_f64.to_radians());
    }

    #[test]
    fn structural_through_arc_solves_a_circle_and_is_identity_free() {
        let arc = through_arc_geometry_kernel(
            coordinate_geometry_kernel(10.0, 0.0),
            coordinate_geometry_kernel(0.0, 10.0),
            coordinate_geometry_kernel(-10.0, 0.0),
            30.0,
            120.0,
        )
        .expect("three non-collinear points define a circle");

        assert!(arc.center.x.abs() < 1e-12);
        assert!(arc.center.y.abs() < 1e-12);
        assert!((arc.radius - 10.0).abs() < 1e-12);
        assert_eq!(arc.start_angle_deg, 30.0);
        assert_eq!(arc.end_angle_deg, 120.0);
        assert_eq!(arc.sweep_angle_deg, 90.0);
        assert_eq!(arc.length, 10.0 * 90.0_f64.to_radians());
        assert!((arc.start.x - 10.0 * 30.0_f64.to_radians().cos()).abs() < 1e-12);
        assert!((arc.end.y - 10.0 * 120.0_f64.to_radians().sin()).abs() < 1e-12);
    }

    #[test]
    fn structural_through_arc_rejects_duplicate_and_collinear_points() {
        let duplicate = through_arc_geometry_kernel(
            coordinate_geometry_kernel(0.0, 0.0),
            coordinate_geometry_kernel(0.0, 0.0),
            coordinate_geometry_kernel(1.0, 1.0),
            0.0,
            90.0,
        );
        let collinear = through_arc_geometry_kernel(
            coordinate_geometry_kernel(0.0, 0.0),
            coordinate_geometry_kernel(1.0, 1.0),
            coordinate_geometry_kernel(2.0, 2.0),
            0.0,
            90.0,
        );

        assert!(duplicate.is_none());
        assert!(collinear.is_none());
    }

    #[test]
    fn structural_polyline_preserves_open_and_closed_geometry() {
        let points = [
            coordinate_geometry_kernel(0.0, 0.0),
            coordinate_geometry_kernel(3.0, 4.0),
            coordinate_geometry_kernel(3.0, 0.0),
        ];
        let open = polyline_geometry_kernel(&points, false).expect("open polyline");
        assert_eq!(open.segments.len(), 2);
        assert_eq!(open.length, 9.0);
        assert_eq!(open.end, points[2]);
        let closed = polyline_geometry_kernel(&points, true).expect("closed polyline");
        assert_eq!(closed.segments.len(), 3);
        assert_eq!(closed.length, 12.0);
        assert_eq!(closed.end, points[0]);
    }
}
